compiler/rustc_codegen_ssa/src/mir/naked_asm.rs

```diff
@@ -104,6 +106,215 @@ fn inline_to_global_operand<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>>(
+
+        Architecture::Aarch64 | Architecture::Aarch64_Ilp32 | Architecture::Arm => {
+            // https://developer.arm.com/documentation/100067/0611/armclang-Integrated-Assembler/AArch32-Target-selection-directives?lang=en
+
+            for feature in features {
+                writeln!(begin, ".arch_extension {}", feature.name).unwrap();
+
+                writeln!(end, ".arch_extension no{}", feature.name).unwrap();
+            }
+        }
+        Architecture::Riscv32 | Architecture::Riscv64 => {
+            // https://github.com/riscv-non-isa/riscv-asm-manual/blob/ad0de8c004e29c9a7ac33cfd054f4d4f9392f2fb/src/asm-manual.adoc#arch
+
+            for feature in features {
+                writeln!(begin, ".option arch, +{}", feature.name).unwrap();
+
+                writeln!(end, ".option arch, -{}", feature.name).unwrap();
+            }
+        }
+        Architecture::Mips | Architecture::Mips64 | Architecture::Mips64_N32 => {
+            // https://sourceware.org/binutils/docs/as/MIPS-ISA.html
+            // https://sourceware.org/binutils/docs/as/MIPS-ASE-Instruction-Generation-Overrides.html
+
+            for feature in features {
+                writeln!(begin, ".set {}", feature.name).unwrap();
+
+                writeln!(end, ".set no{}", feature.name).unwrap();
+            }
+        }
+
+        Architecture::S390x => {
+            // https://sourceware.org/binutils/docs/as/s390-Directives.html
+
+            // based on src/llvm-project/llvm/lib/Target/SystemZ/SystemZFeatures.td
+            let isa_revision_for_feature = |feature: &TargetFeature| match feature.name.as_str() {
+                "backchain" => None, // does not define any instructions
+                "deflate-conversion" => Some(13),
+                "enhanced-sort" => Some(13),
+                "guarded-storage" => Some(12),
+                "high-word" => None, // technically 9, but LLVM supports only >= 10
+                "nnp-assist" => Some(14),
+                "transactional-execution" => Some(10),
+                "vector" => Some(11),
+                "vector-enhancements-1" => Some(12),
+                "vector-enhancements-2" => Some(13),
+                "vector-packed-decimal" => Some(12),
+                "vector-packed-decimal-enhancement" => Some(13),
+                "vector-packed-decimal-enhancement-2" => Some(14),
+                _ => None,
+            };
+
+            if let Some(minimum_isa) = features.filter_map(isa_revision_for_feature).max() {
+                writeln!(begin, ".machine arch{minimum_isa}").unwrap();
+
+                // NOTE: LLVM does not support `.machine push` and `.machine pop`, so we rely on these
+                // target features only being applied to this ASM block (LLVM clears them for the next)
+                //
+                // https://github.com/llvm/llvm-project/blob/74306afe87b85cb9b5734044eb6c74b8290098b3/llvm/lib/Target/SystemZ/AsmParser/SystemZAsmParser.cpp#L1362
+            }
```

Comment:
@uweigand so https://sourceware.org/binutils/docs/as/s390-Directives.html says that `.machine pop` and `.machine push` are supported, but LLVM doesn't as far as I can tell. Is that documentation inaccurate, or is this an oversight on LLVM's part?

LLVM does clear the target features that are in effect for every asm block (from what I can tell), but I don't think e.g. gcc would do that, because it does not actually parse the asm and just passes it on verbatim. So, I'd like to emit `.machine push` and `.machine pop` like powerpc below.

Reply:
Yes, the implementation of directives in the LLVM asm-parser is not complete.   This is mainly because LLVM isn't really used to parse full assembler files, and in inline asm directives are rarely used.  However, in general this should of course be implemented at some point just to be compatible.

Reply:
I opened https://github.com/llvm/llvm-project/issues/129053 so that we have something to link to

Reply:
I had assumed previously that this would not actually be a problem, but it looks like it is: what target features a naked function enables can leak out of an implementation. I've added some logic now that resets the architecture to the minimum required by the global target features, but that is not entirely correct, because users could increase the architecture in global assembly.

I don't think it's likely to come up in practice, but it is inelegant.